sdk/eventhub/Azure.Messaging.EventHubs/tests/Primitives/EventProcessorTests.cs

```diff
@@ -972,6 +1090,540 @@ public async Task StopProcessingLogsFaultedTaskDuringShutdown(bool async)
                 Times.Once);
         }
 
+        /// <summary>
+        ///   Verifies functionality of the <see cref="EventProcessor{TPartition}" />
+        ///   error handler and the <see cref="EventProcessor{TPartition}.StopProcessing" />
+        ///   method.
+        /// </summary>
+        ///
+        [Test]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task StopProcessingIsSafeToCallInTheErrorHandler(bool async)
```

Comment:
The timing for this was tricky; even without the safe-guard loop at the end, it still adds roughly 4 seconds to the local run. _(from ~8.5 seconds to ~12.4 seconds)_    That said, it proved valuable as it caught a deadlock in what should be a valid scenario, so I definitely do not want to remove it.

I'd be grateful for any thoughts on how to potentially tune it or better approaches that allow for deterministic validation.

Reply:
Instead of setting the result of the `completionSource` in the `StopProcessing` callback, couldn't we set it when calling `OnProcessingErrorAsync`, after the try/catch block? This way we can make sure the processor has stopped before the verification. I don't think this will be very helpful in improving the total time, though.

Reply:
I could have sworn that I had it that way at one point and moved it over for some reason.   That said, it does make sense to have it fire after `OnProcessingErrorAsync` has returned, since that *should* ensure that the Stop call has fully completed.    Looks like the execution time remains the same, sadly.

I'm going to leave the safety loop in there just to guard against non-determinism, since the `Status` property intentionally does not attempt to acquire the semaphore when the processor is stopping, there is still a small (but acceptably small) chance that it could see `Stopping` rather than `Stopped` due grabbing a copy of the override value at exactly the wrong moment.

Reply:
Random idea -- If you provide a mock implementation for `Logger` the completionSource could be set on the callback to `EventProcessorStopComplete` instead, which happens after the `Status` is stable.

Reply:
Clever idea!  Gave it a try and it looks stable and deterministic in local runs without the safety loop.   Bringing it forward into these changes and will adjust if we see anything weird in nightly or CI runs.